src/Storages/MergeTree/ReplicatedMergeMutateTaskBase.cpp

```diff
@@ -0,0 +1,225 @@
+                auto mutations_end_it = in_partition->second.upper_bound(result_data_version);
+                for (auto it = mutations_begin_it; it != mutations_end_it; ++it)
+                {
+                    ReplicatedMergeTreeQueue::MutationStatus & status = *it->second;
+                    status.latest_failed_part = log_entry->source_parts.at(0);
+                    status.latest_failed_part_info = source_part_info;
+                    status.latest_fail_time = time(nullptr);
+                    status.latest_fail_reason = getExceptionMessage(saved_exception, false);
+                }
+            }
+        }
+
+    }
+
+    return false;
+}
+
+
+bool ReplicatedMergeMutateTaskBase::executeImpl()
+{
+    MemoryTrackerThreadSwitcherPtr switcher;
+    if (merge_mutate_entry)
+        switcher = std::make_unique<MemoryTrackerThreadSwitcher>(&(*merge_mutate_entry)->memory_tracker);
+
+    switch (state)
+    {
+        case State::NEED_PREPARE :
+        {
+            prepareCommon();
+
+            /// Depending on condition there is no need to execute a merge
+            if (state == State::SUCCESS)
+                return true;
+
+            if (!prepare())
+            {
+                state = State::CANT_MERGE_NEED_FETCH;
+                return true;
+            }
+
+            state = State::NEED_EXECUTE_INNER_MERGE;
+            return true;
+        }
+        case State::NEED_EXECUTE_INNER_MERGE :
+        {
+            try
+            {
+                if (!executeInnerTask())
+                {
+                    state = State::NEED_COMMIT;
+                    return true;
+                }
+            }
+            catch (...)
+            {
+                /// Maybe part is nullptr here?
+                write_part_log(ExecutionStatus::fromCurrentException());
+                throw;
+            }
+
+            return true;
+        }
+        case State::NEED_COMMIT :
+        {
+            try
+            {
+                if (!finalize())
+                {
+                    state = State::CANT_MERGE_NEED_FETCH;
+                    return true;
+                }
+            }
+            catch (...)
+            {
+                write_part_log(ExecutionStatus::fromCurrentException());
+                throw;
+            }
+
+            state = State::SUCCESS;
+            return true;
+        }
+        case State::CANT_MERGE_NEED_FETCH :
+        {
+            if (storage.executeFetch(entry))
+            {
+                state = State::SUCCESS;
+                return true;
+            }
+
+            return false;
+        }
+        case State::SUCCESS :
+        {
+            storage.queue.removeProcessedEntry(storage.getZooKeeper(), selected_entry->log_entry);
+            return false;
+        }
+    }
+    return false;
+}
+
+
+void ReplicatedMergeMutateTaskBase::prepareCommon()
```

Comment:
`bool checkExistingPart()`?  
It would be much easier if status is only changed in `executeImpl`

Reply:
Would be, but current code is so. I don't want to change a logic, only move some code from one place to another.

Reply:
Fixed